### DISCUSSION TITLE
Using Thebe to run code on CoCalc?

### DISCUSSION THREAD

**Discussion by mforbes** [sagemathinc/cocalc#7145] [Ideas]
How hard would it be to configure and use [Thebe](https://github.com/executablebooks/thebe) or something similar to allow code hosted on either CoCalc or ReadTheDocs etc. to launch and execute on CoCalc instead of MyBinder?

I am just starting to explore this, so don't know what all is required.  Has anyone thought about something similar?  (I would really like to be able to host [Jupyter Books](https://jupyterbook.org/en/stable/intro.html), not just notebooks.)

Currently, there are lots of things hard-coded (i.e. repo must be hosted on GitHub, kernel information etc.) but it seems like it should not be too hard to update/change as needed.

# References
* [Thebe documentation](https://thebe.readthedocs.io/en/stable/)
* [sphinx-thebe extension documentation](https://sphinx-thebe.readthedocs.io/en/latest/)
* [Jupyter Book interactive code example](https://jupyterbook.org/en/stable/interactive/thebe.html)

**Comment by haraldschilly** [comment]
I don't know thebe. CoCalc already allows to execute code blocks embedded in markdown or just straight a published jupyter notebook, when shared on the public server → [example](https://cocalc.com/haraldschilly/ideas/sagemath-in-markdown).

Maybe it is possible to run generated markdown of Jupyter Books directly? (or if not, to come up with a plugin for it, to make it work?)

Technically, it is possible to run any kind of server on your own using a [compute server](https://doc.cocalc.com/compute_server.html), though.

**Reply by mforbes** [reply]
The idea is a little different than what CoCalc does for published notebooks in that one would need to spawn a kernel (like mybinder does) where the cells in the notebook/markdown file could be executed.  It is not enough to just push each cell - the intermediate state needs to be maintained so users can play.  (I suppose that the entire notebook could be re-executed on code-changes, which would then fit the current CoCalc model, but that would probably not be so good from a user's perspective if the notebook contained substantial code.)

What I would like to be able to do is to be able to use one's CoCalc license to run a kernel with more resources than mybinder, for example.  Of course, this could be done by having users install the jupyterbook in a project, then run it with a compute server as you describe, but I would like to try to automate this as much as possible so that students, for example, can just click a button on the docs, sign into CoCalc, and then play with the code.

I am also thinking of this as a good way to attract people to CoCalc.

Ultimately, once CoCalc [supports Myst stuff in the Markdown editor](https://github.com/sagemathinc/cocalc/issues/4715), it will probably be possible to get a good rendition there (though there is something appealing to having users able to play with code without changing the documentation parts for education purposes.)
